### PR TITLE
All divulgence events pruning

### PR DESCRIPTION
**Pull request by tudor-da**
All divulgence events pruning
* Add `prune_all_divulged_contracts` in PruneRequest
* Add `participant_pruned_all_contracts_up_to_inclusive` in parameters table
* pruneEvents prunes all divulgence events

CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/participant-integration-api/src/main/scala/platform/store/backend/common/CommonStorageBackend.scala [hunk 1]
```diff
@@ -606,18 +616,56 @@ private[backend] trait CommonStorageBackend[DB_BATCH] extends StorageBackend[DB_
               WHERE
                 archive_events.event_offset <= $pruneUpToInclusive AND
                 archive_events.contract_id = delete_events.contract_id
-            )""",
+            )"""
+    }(connection, loggingContext)
+
+    pruneWithLogging(queryDescription = "Exercise (consuming) events pruning") {
```
[the remark this answers is not in the document]
Comment (meiersi-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/backend/common/CommonStorageBackend.scala:622]: Also, if this bug was not caught by a test. Consider, before fixing it, what test you can add that would have caught it.

[the remark this answers is not in the document]
Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/backend/common/CommonStorageBackend.scala:622]: Yes, you're right, thanks! When not pruning all divulgence events, this will leave orphans. Oversaw this. Will amend. Wrt to the test, the new tests are only asserting prune with `prune_all_divulged_contracts`.

File: ledger/participant-integration-api/src/main/scala/platform/store/backend/common/CommonStorageBackend.scala [hunk 2]
```diff
@@ -606,18 +616,56 @@ private[backend] trait CommonStorageBackend[DB_BATCH] extends StorageBackend[DB_
...
+      }(connection, loggingContext)
+    }
+  }
+
+  private def pruneWithLogging(queryDescription: String)(query: SimpleSql[Row])(
+      connection: Connection,
+      loggingContext: LoggingContext,
+  ): Unit = {
+    val deletedRows = query.executeUpdate()(connection)
+    logger.info(s"$queryDescription finished: deleted $deletedRows rows.")(loggingContext)
```
[the remark this answers is not in the document]
Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/backend/common/CommonStorageBackend.scala:671]: Done.

File: ledger/participant-integration-api/src/main/scala/platform/store/backend/common/CommonStorageBackend.scala [hunk 3]
```diff
@@ -606,18 +616,56 @@ private[backend] trait CommonStorageBackend[DB_BATCH] extends StorageBackend[DB_
...
           -- Exercise events (non-consuming)
           delete from participant_events_non_consuming_exercise delete_events
           where
-            delete_events.event_offset <= $pruneUpToInclusive""",
-    ).foreach(_.execute()(connection))
+            delete_events.event_offset <= $pruneUpToInclusive"""
+    }(connection, loggingContext)
+
+    if (pruneAllDivulgedContracts) {
+      pruneWithLogging(queryDescription = "Retroactive divulgence events pruning") {
```
[the remark this answers is not in the document]
Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/backend/common/CommonStorageBackend.scala]: Done

File: ledger/participant-integration-api/src/main/scala/platform/store/backend/common/CommonStorageBackend.scala [hunk 4]
```diff
@@ -606,18 +616,56 @@ private[backend] trait CommonStorageBackend[DB_BATCH] extends StorageBackend[DB_
...
+        SQL"""
+          -- Retroactive divulgence events
+          delete from participant_events_divulgence delete_events
+          where delete_events.event_offset <= $pruneUpToInclusive
+          """
+      }(connection, loggingContext)
+    } else {
+      pruneWithLogging(queryDescription = "Retroactive divulgence events pruning") {
```
[the remark this answers is not in the document]
Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/backend/common/CommonStorageBackend.scala]: Done

File: ledger/participant-integration-api/src/main/scala/platform/store/backend/common/CommonStorageBackend.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/backend/common/CommonStorageBackend.scala:649]: Added a new test in `ParticipantPruningIT` (`PRDivulgenceArchivalPruning`) to fill this gap.

**Review by meiersi-da — APPROVED** [review]
LGTM

**Review by hubert-da — COMMENTED** [review]
LGTM from the code perspective 👍🏻 don't have enough context to approve

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ParticipantPruningIT.scala [hunk 5]
```diff
@@ -531,6 +531,58 @@ class ParticipantPruningIT extends LedgerTestSuite {
...
+
+      offsetAfterDivulgence_2 <- beta.currentEnd()
+
+      _ <- alpha.exercise(alice, contract.exerciseArchive)
+
+      _ <- synchronize(alpha, beta)
+
+      _ <- beta.prune(offsetAfterDivulgence_2, pruneAllDivulgedContracts = false)
+
+      // Bob cannot see the divulged contract
```
Comment (mziolekda) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ParticipantPruningIT.scala:576]: Should we not check the way in which this fails? 
Since we have not pruned all the divulged contracts, this call will succeed synchronously but fail on the completion stream. This would be different if we pruned the divulged contracts - there the error would be reported synchronously.

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/SpannedIndexService.scala [hunk 6]
```diff
@@ -196,10 +196,10 @@ private[daml] final class SpannedIndexService(delegate: IndexService) extends In
   )(implicit loggingContext: LoggingContext): Future[Unit] =
     delegate.stopDeduplicatingCommand(commandId, submitter)
 
-  override def prune(pruneUpToInclusive: Offset)(implicit
+  override def prune(pruneUpToInclusive: Offset, pruneAllDivulgedContracts: Boolean)(implicit
       loggingContext: LoggingContext
   ): Future[Unit] =
-    delegate.prune(pruneUpToInclusive)
+    delegate.prune(pruneUpToInclusive, pruneAllDivulgedContracts: Boolean)
```
Comment (mziolekda) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/SpannedIndexService.scala:202]:
```suggestion
    delegate.prune(pruneUpToInclusive, pruneAllDivulgedContracts)
```

File: ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/JdbcLedgerDao.scala [hunk 7]
```diff
@@ -604,14 +597,36 @@ private class JdbcLedgerDao(
...
+          storageBackend.updatePrunedAllDivulgedContractsUpToInclusive(pruneUpToInclusive)(conn)
+        }
+      }
+      .andThen {
+        case Success(_) =>
+          logger.info(
+            s"Completed pruning of the ledger api server index db$allDivulgencePruningParticle up to ${pruneUpToInclusive.toHexString}"
+          )
+        case Failure(ex) =>
+          logger.warn("Pruning failed", ex)
```
Comment (mziolekda) [inline ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/JdbcLedgerDao.scala:627]: I would change the logging around pruning somewhat:
- include the pruneUpToInclusive in the first log at the entry to the function
- Rephrase the initial messge to: "Pruning the ledger api server index db$allDivulgencePruningParticle up to ${pruneUpToInclusive.toHexString}"
- Shorten the successful completion to just "Pruning succeeded"

**Review by mziolekda — APPROVED** [review]
I have left a couple of minor comments

General question: will you handle the corner cases described in the story ACs in a separate PR: 
```
when migrating index databases that contain divulged events that were created under 
a) mutating schema 
b) append-only schema with non transaction-local
```

**Comment by meiersi-da** [comment]
@mziolekda : these corner cases required documentation only don't they?

What is though true is that this PR misses the check that the `pruning_offset` is after the migration-to-append-only offset in case `prune_all_divulged_contracts` is set.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ParticipantPruningIT.scala (hunk 5 above)
Comment (tudor-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ParticipantPruningIT.scala:576]: The purpose of this test is to show that divulgence events are pruned if their associated contract is archived. So the failure happens synchronously. 

I see your point about making sure that the failure is happening during command submission and not due to missing the tree events on the output stream. But I don't see how can this submission pass and the events stream to succeed. In case I understood wrongly, can you clarify?

File: ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/JdbcLedgerDao.scala (hunk 7 above)
Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/JdbcLedgerDao.scala:629]: Done

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/SpannedIndexService.scala (hunk 6 above)
Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/SpannedIndexService.scala]: Done

**Comment by tudor-da** [comment]
> @mziolekda : these corner cases require documentation only don't they?
> 
> What is though true is that this PR misses the check that the `pruning_offset` is after the migration-to-append-only offset in case `prune_all_divulged_contracts` is set.

Actually, I want to leave the check out of this PR:
* The current PR is quite comprehensive
* The existing logic enforces pruning only after migration implicitly (pre-migration NULL values don't have offsets)
* The solution for enforcing this check is not yet clear and might be complex

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ParticipantPruningIT.scala (hunk 5 above)
Comment (tudor-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ParticipantPruningIT.scala]: On the other hand, thanks for drawing my attention again to this test. In the original form, it could not pass for privacy-aware ledgers. Changed it to a single-participant test, so it can work on privacy-aware ledgers as well.

**Comment by tudor-da** [comment]
> @mziolekda : these corner cases require documentation only don't they?
> 
> What is though true is that this PR misses the check that the `pruning_offset` is after the migration-to-append-only offset in case `prune_all_divulged_contracts` is set.

Not sure which one would be the place to document the pruning requirements and corner cases, since it involves topics touching different components. The best place for sure is in the SDK docs, but for having it in code as well, I've added some brief remarks over `appendonlydao.JdbcLedgerDao.prune`

File: ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/JdbcLedgerDao.scala [hunk 8]
```diff
@@ -598,15 +592,58 @@ private class JdbcLedgerDao(
...
+    *
+    * @param pruneUpToInclusive         Offset up to which to prune archived history inclusively.
+    * @param pruneAllDivulgedContracts  Enables pruning of all immediately and retroactively divulged contracts
+    *                                   up to `pruneUpToInclusive`.
+    *
+    * NOTE:  Pruning of all divulgence events needs to happen taking into account the following considerations:
+    *        1.  Pruning of all divulged contracts needs to use `pruneUpToInclusive`
+    *            beyond `participant_all_divulged_contracts_pruned_up_to_inclusive`
+    *            (i.e. the last successfully used `pruneUpToInclusive` with `pruneAllDivulgedContracts` set).
+    *        2.  Pruning of all divulged contracts needs to use `pruneUpToInclusive` past the last offset ingested
```
[the remark this answers is not in the document]
Comment (meiersi-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/JdbcLedgerDao.scala:605]: consider adding a "consideration-name:" for every consideration

File: ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/JdbcLedgerDao.scala [hunk 9]
```diff
@@ -598,15 +592,58 @@ private class JdbcLedgerDao(
...
+    *            before the migration to append-only schema (if there is one)
+    *            since the divulgence events prior to that point do not have an offset.
+    *        3.  Ledgers populated with WriteService versions that do not forward transaction-local divulgence
+    *            will hydrate the index with the divulgence events only once for a specific contract-party divulgence relationship
+    *            regardless of the number of re-divulgences of the contract to the same party have occurred after the initial one.
+    *            In this case, pruning of all divulged contracts might hinder visibility even for submissions
```
Comment (meiersi-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/JdbcLedgerDao.scala:611]:
```suggestion
    *            In this case, pruning of all divulged contracts might lead to interpretation failures for command submissions
    *            despite them <remainder of your text>
```

**Review by meiersi-da — COMMENTED** [review]
@tudor-da : I've done a quick review of your code docs. I hope my input on how to improve clarity for future team members reading them is welcome. Given the subtlety of the divulgence topic, I felt that it would be well invested time to polish these docs for clarity.

File: ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/JdbcLedgerDao.scala [hunk 10]
```diff
@@ -592,6 +592,25 @@ private class JdbcLedgerDao(
...
+    *            which rely on divulgences that happened after the `pruneUpToInclusive` offset.
+    */
```
[the remark this answers is not in the document]
Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/JdbcLedgerDao.scala:613]: Ok, will enrich the docs here then 👍

File: ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/JdbcLedgerDao.scala [hunk 11]
```diff
@@ -598,15 +592,58 @@ private class JdbcLedgerDao(
...
+  /** Prunes the events and command completions tables.
+    *
+    * @param pruneUpToInclusive         Offset up to which to prune archived history inclusively.
+    * @param pruneAllDivulgedContracts  Enables pruning of all immediately and retroactively divulged contracts
+    *                                   up to `pruneUpToInclusive`.
+    *
+    * NOTE:  Pruning of all divulgence events needs to happen taking into account the following considerations:
+    *        1.  Pruning of all divulged contracts needs to use `pruneUpToInclusive`
+    *            beyond `participant_all_divulged_contracts_pruned_up_to_inclusive`
+    *            (i.e. the last successfully used `pruneUpToInclusive` with `pruneAllDivulgedContracts` set).
```
[the remark this answers is not in the document]
Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/JdbcLedgerDao.scala:604]: Exactly. However, I realized there's a better place for this remark since it is not a constraint on the input arguments but just an info on how pruning will work for disclosures. Will remove from here and add with the disclosure divulgence PR.

File: ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/JdbcLedgerDao.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/JdbcLedgerDao.scala:605]: Done, made the docs a bit more involved and added a consideration tag

File: ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/JdbcLedgerDao.scala (hunk 9 above)
Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/JdbcLedgerDao.scala:611]: Done

File: ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/JdbcLedgerDao.scala [hunk 12]
```diff
@@ -598,15 +592,58 @@ private class JdbcLedgerDao(
     )
   }
 
+  /** Prunes the events and command completions tables.
+    *
+    * @param pruneUpToInclusive         Offset up to which to prune archived history inclusively.
+    * @param pruneAllDivulgedContracts  Enables pruning of all immediately and retroactively divulged contracts
+    *                                   up to `pruneUpToInclusive`.
+    *
+    * NOTE:  Pruning of all divulgence events needs to happen taking into account the following considerations:
```
[the remark this answers is not in the document]
Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/JdbcLedgerDao.scala]: Done

File: ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/JdbcLedgerDao.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/JdbcLedgerDao.scala]: Let me know if you consider something else missing.
